fix: prevent floats without decimal values from being encoded as int

kruskall opened the pull request:
<!-- Thanks for sending a pull request!

If this is your first contribution, please review and sign our contributor agreement -
https://www.elastic.co/contributor-agreement.

Guidelines:
 - Prefer small PRs, and split changes into multiple logical commits where they must
   be delivered in a single PR.
 - If the PR is incomplete and not yet ready for review, open it as a Draft.
 - Once the PR is marked ready for review it is expected to pass all tests and linting,
   and you should not force-push any changes.

See also https://github.com/elastic/apm-server/blob/main/CONTRIBUTING.md for more tips on contributing.
-->

## Motivation/summary

Floats are encoded as ints and trailing zeroes are being stripped,
causing document rejection due to mismatched types.
APM server is relying on fastjson, which is using strconv, to encode
metrics.
The precision is set to -1 so that the encoded string uses the
smallest number of digits. To avoid losing type information, truncate
the value by casting to int and compare the result to the original
value: if they are equal conversion is carried out with precision=1
so that the trailing zero is not stripped.

## Checklist

<!--
Delete irrelevant items. The changelog should only be updated for user-facing changes.
Once the PR is ready for review there should be no unticked boxes.
-->

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

<!--
Explain how this PR can be tested by the reviewer: commands, dependencies, steps, etc.
If it is self-explanatory, delete this section.
-->

## Related issues

Closes https://github.com/elastic/apm-server/issues/5682

apmmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8868/detail/PR-8868/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8868/detail/PR-8868/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8868/detail/PR-8868/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8868/detail/PR-8868/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-server_8868.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-08-15T09:28:45.417Z&rangeTo=2022-08-15T09:48:45.417Z&transactionName=BUILD+apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=0d36438360c63d77732fdb18ed607bcc&transactionId=0687c6bb66aa9b94)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-08-15T09:38:45.417+0000


* Duration: 22 min 27 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 126  |
| Skipped      | 1 |
| Total        | 128   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8868/detail/PR-8868/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `TestFleetIntegrationMonitoring – github.com/elastic/apm-server/systemtest`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestFleetIntegrationMonitoring
2022/08/15 09:58:14 Starting container id: 3b2986e07bab image: elastic-agent-systemtest:8.5.0-60a4c029-SNAPSHOT
2022/08/15 09:58:15 Waiting for container id 3b2986e07bab image: elastic-agent-systemtest:8.5.0-60a4c029-SNAPSHOT
2022/08/15 09:58:20 Container is ready id: 3b2986e07bab image: elastic-agent-systemtest:8.5.0-60a4c029-SNAPSHOT
    fleet_test.go:88: 
        	Error Trace:	/var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-8868/src/github.com/elastic/apm-server/systemtest/fleet_test.go:88
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"output":map[string]interface {}{"events":map[string]interface {}{"acked":15, "active":0, "batches":1, "failed":0, "toomany":0, "total":15}, "type":"elasticsearch", "write":map[string]interface {}{}}, "pipeline":map[string]interface {}{"events":map[string]interface {}{"total":15}}}
        	            	actual  : map[string]interface {}{"config":map[string]interface {}{"module":map[string]interface {}{"running":0, "starts":0, "stops":0}, "reloads":0, "scans":0}, "output":map[string]interface {}{"events":map[string]interface {}{"acked":0, "active":0, "batches":0, "dropped":0, "duplicates":0, "failed":0, "toomany":0, "total":0}, "read":map[string]interface {}{"bytes":0, "errors":0}, "type":"elasticsearch", "write":map[string]interface {}{"errors":0}}, "pipeline":map[string]interface {}{"clients":0, "events":map[string]interface {}{"active":0, "dropped":0, "failed":0, "filtered":0, "published":0, "retry":0, "total":0}, "queue":map[string]interface {}{"acked":0, "max_events":4096}}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,18 +1,45 @@
        	            	-(map[string]interface {}) (len=2) {
        	            	- (string) (len=6) "output": (map[string]interface {}) (len=3) {
        	            	-  (string) (len=6) "events": (map[string]interface {}) (len=6) {
        	            	-   (string) (len=5) "acked": (float64) 15,
        	            	+(map[string]interface {}) (len=3) {
        	            	+ (string) (len=6) "config": (map[string]interface {}) (len=3) {
        	            	+  (string) (len=6) "module": (map[string]interface {}) (len=3) {
        	            	+   (string) (len=7) "running": (float64) 0,
        	            	+   (string) (len=6) "starts": (float64) 0,
        	            	+   (string) (len=5) "stops": (float64) 0
        	            	+  },
        	            	+  (string) (len=7) "reloads": (float64) 0,
        	            	+  (string) (len=5) "scans": (float64) 0
        	            	+ },
        	            	+ (string) (len=6) "output": (map[string]interface {}) (len=4) {
        	            	+  (string) (len=6) "events": (map[string]interface {}) (len=8) {
        	            	+   (string) (len=5) "acked": (float64) 0,
        	            	    (string) (len=6) "active": (float64) 0,
        	            	-   (string) (len=7) "batches": (float64) 1,
        	            	+   (string) (len=7) "batches": (float64) 0,
        	            	+   (string) (len=7) "dropped": (float64) 0,
        	            	+   (string) (len=10) "duplicates": (float64) 0,
        	            	    (string) (len=6) "failed": (float64) 0,
        	            	    (string) (len=7) "toomany": (float64) 0,
        	            	-   (string) (len=5) "total": (float64) 15
        	            	+   (string) (len=5) "total": (float64) 0
        	            	+  },
        	            	+  (string) (len=4) "read": (map[string]interface {}) (len=2) {
        	            	+   (string) (len=5) "bytes": (float64) 0,
        	            	+   (string) (len=6) "errors": (float64) 0
        	            	   },
        	            	   (string) (len=4) "type": (string) (len=13) "elasticsearch",
        	            	-  (string) (len=5) "write": (map[string]interface {}) {
        	            	+  (string) (len=5) "write": (map[string]interface {}) (len=1) {
        	            	+   (string) (len=6) "errors": (float64) 0
        	            	   }
        	            	  },
        	            	- (string) (len=8) "pipeline": (map[string]interface {}) (len=1) {
        	            	-  (string) (len=6) "events": (map[string]interface {}) (len=1) {
        	            	-   (string) (len=5) "total": (float64) 15
        	            	+ (string) (len=8) "pipeline": (map[string]interface {}) (len=3) {
        	            	+  (string) (len=7) "clients": (float64) 0,
        	            	+  (string) (len=6) "events": (map[string]interface {}) (len=7) {
        	            	+   (string) (len=6) "active": (float64) 0,
        	            	+   (string) (len=7) "dropped": (float64) 0,
        	            	+   (string) (len=6) "failed": (float64) 0,
        	            	+   (string) (len=8) "filtered": (float64) 0,
        	            	+   (string) (len=9) "published": (float64) 0,
        	            	+   (string) (len=5) "retry": (float64) 0,
        	            	+   (string) (len=5) "total": (float64) 0
        	            	+  },
        	            	+  (string) (len=5) "queue": (map[string]interface {}) (len=2) {
        	            	+   (string) (len=5) "acked": (float64) 0,
        	            	+   (string) (len=10) "max_events": (float64) 4096
        	            	   }
        	Test:       	TestFleetIntegrationMonitoring
    fleet_test.go:175: elastic-agent logs: {"log.level":"info","@timestamp":"2022-08-15T09:58:16.432Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":471},"message":"Starting enrollment to URL: https://fleet-server:8220/","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.217Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":271},"message":"Elastic Agent might not be running; unable to trigger restart","ecs.version":"1.6.0"}
        Successfully enrolled the Elastic Agent.
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.366Z","log.origin":{"file.name":"cmd/run.go","file.line":202},"message":"APM instrumentation disabled","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.367Z","log.origin":{"file.name":"application/application.go","file.line":78},"message":"Detecting execution mode","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.385Z","log.origin":{"file.name":"application/application.go","file.line":102},"message":"Agent is managed by Fleet","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.385Z","log.origin":{"file.name":"capabilities/capabilities.go","file.line":59},"message":"capabilities file not found in /usr/share/elastic-agent/state/capabilities.yml","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.863Z","log.logger":"composable.providers.docker","log.origin":{"file.name":"docker/docker.go","file.line":44},"message":"Docker provider skipped, unable to connect: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.864Z","log.logger":"api","log.origin":{"file.name":"api/server.go","file.line":68},"message":"Starting stats endpoint","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.864Z","log.origin":{"file.name":"application/managed_mode.go","file.line":323},"message":"Agent is starting","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:17.864Z","log.logger":"api","log.origin":{"file.name":"api/server.go","file.line":70},"message":"Metrics endpoint listening on: /usr/share/elastic-agent/state/data/tmp/elastic-agent.sock (configured: unix:///usr/share/elastic-agent/state/data/tmp/elastic-agent.sock)","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:19.539Z","log.origin":{"file.name":"artifact/config.go","file.line":138},"message":"Source URI changed from \"https://artifacts.elastic.co/downloads/\" to \"https://artifacts.elastic.co/downloads/\"","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:19.606Z","log.origin":{"file.name":"stateresolver/stateresolver.go","file.line":48},"message":"New State ID is B5tgjcxp","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:19.606Z","log.origin":{"file.name":"stateresolver/stateresolver.go","file.line":49},"message":"Converging state requires execution of 2 step(s)","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:19.787Z","log.origin":{"file.name":"operation/operator.go","file.line":348},"message":"operation 'operation-install' skipped for apm-server.8.5.0-SNAPSHOT","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:19.913Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-08-15T09:58:19Z - message: Application: apm-server--8.5.0-SNAPSHOT[08803aa2-23c9-4a03-99e6-5bed2aab0cbe]: State changed to STARTING: Starting - type: 'STATE' - sub_type: 'STARTING'","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:20.247Z","log.origin":{"file.name":"operation/operator.go","file.line":348},"message":"operation 'operation-install' skipped for filebeat.8.5.0-SNAPSHOT","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:20.532Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-08-15T09:58:20Z - message: Application: filebeat--8.5.0-SNAPSHOT--36643631373035623733363936343635[08803aa2-23c9-4a03-99e6-5bed2aab0cbe]: State changed to STARTING: Starting - type: 'STATE' - sub_type: 'STARTING'","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:21.004Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-08-15T09:58:21Z - message: Application: apm-server--8.5.0-SNAPSHOT[08803aa2-23c9-4a03-99e6-5bed2aab0cbe]: State changed to RUNNING: Running - type: 'STATE' - sub_type: 'RUNNING'","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:21.028Z","log.origin":{"file.name":"operation/operator.go","file.line":348},"message":"operation 'operation-install' skipped for metricbeat.8.5.0-SNAPSHOT","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:21.181Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-08-15T09:58:21Z - message: Application: metricbeat--8.5.0-SNAPSHOT--36643631373035623733363936343635[08803aa2-23c9-4a03-99e6-5bed2aab0cbe]: State changed to STARTING: Starting - type: 'STATE' - sub_type: 'STARTING'","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:21.184Z","log.origin":{"file.name":"stateresolver/stateresolver.go","file.line":66},"message":"Updating internal state","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:21.834Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-08-15T09:58:21Z - message: Application: filebeat--8.5.0-SNAPSHOT--36643631373035623733363936343635[08803aa2-23c9-4a03-99e6-5bed2aab0cbe]: State changed to RUNNING: Running - type: 'STATE' - sub_type: 'RUNNING'","ecs.version":"1.6.0"}
        {"log.level":"info","@timestamp":"2022-08-15T09:58:22.508Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-08-15T09:58:22Z - message: Application: metricbeat--8.5.0-SNAPSHOT--36643631373035623733363936343635[08803aa2-23c9-4a03-99e6-5bed2aab0cbe]: State changed to RUNNING: Running - type: 'STATE' - sub_type: 'RUNNING'","ecs.version":"1.6.0"}
    fleet_test.go:177: apm-server logs:
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.974Z","log.origin":{"file.name":"instance/beat.go","file.line":710},"message":"Beat ID: 62da2c3a-ed6f-4e2e-85c5-a03a56cf45bb","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.974Z","log.origin":{"file.name":"instance/beat.go","file.line":727},"message":"Set gc percentage to: 100","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.975Z","log.logger":"api","log.origin":{"file.name":"api/server.go","file.line":63},"message":"Starting stats endpoint","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.975Z","log.logger":"beat","log.origin":{"file.name":"instance/beat.go","file.line":1076},"message":"Beat info","service.name":"apm-server","system_info":{"beat":{"path":{"config":"/usr/share/elastic-agent/data/elastic-agent-448b21/install/apm-server-8.5.0-SNAPSHOT-linux-x86_64","data":"/usr/share/elastic-agent/state/data/run/default/apm-server--8.5.0-SNAPSHOT","home":"/usr/share/elastic-agent/data/elastic-agent-448b21/install/apm-server-8.5.0-SNAPSHOT-linux-x86_64","logs":"/usr/share/elastic-agent/data/elastic-agent-448b21/install/apm-server-8.5.0-SNAPSHOT-linux-x86_64/logs"},"type":"apm-server","uuid":"62da2c3a-ed6f-4e2e-85c5-a03a56cf45bb"},"ecs.version":"1.6.0"}}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.975Z","log.logger":"beat","log.origin":{"file.name":"instance/beat.go","file.line":1085},"message":"Build info","service.name":"apm-server","system_info":{"build":{"commit":"82f8ec0d4292f361130fc953ed8a5e7bfbe12e1c","libbeat":"8.5.0","time":"2022-08-15T11:30:14.000+0200","version":"8.5.0"},"ecs.version":"1.6.0"}}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.975Z","log.logger":"beat","log.origin":{"file.name":"instance/beat.go","file.line":1088},"message":"Go runtime info","service.name":"apm-server","system_info":{"go":{"os":"linux","arch":"amd64","max_procs":4,"version":"go1.17.12"},"ecs.version":"1.6.0"}}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.976Z","log.logger":"beat","log.origin":{"file.name":"instance/beat.go","file.line":1092},"message":"Host info","service.name":"apm-server","system_info":{"host":{"architecture":"x86_64","boot_time":"2022-08-15T09:47:34Z","containerized":true,"name":"3b2986e07bab","ip":["127.0.0.1/8","172.18.0.7/16"],"kernel_version":"5.4.0-1084-gcp","mac":["02:42:ac:12:00:07"],"os":{"type":"linux","family":"debian","platform":"ubuntu","name":"Ubuntu","version":"20.04.4 LTS (Focal Fossa)","major":20,"minor":4,"patch":4,"codename":"focal"},"timezone":"UTC","timezone_offset_sec":0},"ecs.version":"1.6.0"}}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.976Z","log.logger":"beat","log.origin":{"file.name":"instance/beat.go","file.line":1121},"message":"Process info","service.name":"apm-server","system_info":{"process":{"capabilities":{"inheritable":null,"permitted":null,"effective":null,"bounding":["chown","dac_override","fowner","fsetid","kill","setgid","setuid","setpcap","net_bind_service","net_raw","sys_chroot","mknod","audit_write","setfcap"],"ambient":null},"cwd":"/usr/share/elastic-agent/data/elastic-agent-448b21/install/apm-server-8.5.0-SNAPSHOT-linux-x86_64","exe":"/usr/share/elastic-agent/data/elastic-agent-448b21/install/apm-server-8.5.0-SNAPSHOT-linux-x86_64/apm-server","name":"apm-server","pid":25,"ppid":6,"seccomp":{"mode":"filter","no_new_privs":false},"start_time":"2022-08-15T09:58:19.430Z"},"ecs.version":"1.6.0"}}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.976Z","log.origin":{"file.name":"instance/beat.go","file.line":293},"message":"Setup Beat: apm-server; Version: 8.5.0","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.979Z","log.logger":"esclientleg","log.origin":{"file.name":"eslegclient/connection.go","file.line":108},"message":"elasticsearch url: http://localhost:9200","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.980Z","log.logger":"publisher","log.origin":{"file.name":"pipeline/module.go","file.line":113},"message":"Beat name: 3b2986e07bab","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.981Z","log.logger":"api","log.origin":{"file.name":"api/server.go","file.line":65},"message":"Metrics endpoint listening on: /usr/share/elastic-agent/state/data/tmp/default/apm-server/apm-server.sock (configured: unix:///usr/share/elastic-agent/state/data/tmp/default/apm-server/apm-server.sock)","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.989Z","log.origin":{"file.name":"instance/beat.go","file.line":470},"message":"apm-server start running.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2022-08-15T09:58:19.989Z","log.logger":"cfgwarn","log.origin":{"file.name":"management/manager.go","file.line":116},"message":"BETA: Fleet management is enabled","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.989Z","log.logger":"centralmgmt.fleet","log.origin":{"file.name":"management/manager.go","file.line":117},"message":"Starting fleet management service","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.990Z","log.logger":"centralmgmt.fleet","log.origin":{"file.name":"management/manager.go","file.line":125},"message":"Ready to receive configuration","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:19.991Z","log.logger":"beater","log.origin":{"file.name":"beater/beater.go","file.line":1058},"message":"maxprocs: Leaving GOMAXPROCS=4: CPU quota undefined","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.505Z","log.logger":"centralmgmt.fleet","log.origin":{"file.name":"management/manager.go","file.line":162},"message":"Status change to Configuring: Updating configuration","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.507Z","log.logger":"centralmgmt.fleet","log.origin":{"file.name":"management/manager.go","file.line":287},"message":"Applying settings for inputs","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2022-08-15T09:58:20.507Z","log.logger":"cfgwarn","log.origin":{"file.name":"tlscommon/config.go","file.line":102},"message":"DEPRECATED: Treating the CommonName field on X.509 certificates as a host name when no Subject Alternative Names are present is going to be removed. Please update your certificates if needed. Will be removed in version: 8.0.0","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.508Z","log.logger":"config","log.origin":{"file.name":"config/rum.go","file.line":86},"message":"Unable to determine sourcemap storage, sourcemaps will not be applied","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.508Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":144},"message":"Listening on: [::]:8200","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.509Z","log.logger":"centralmgmt.fleet","log.origin":{"file.name":"management/manager.go","file.line":287},"message":"Applying settings for output","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.509Z","log.logger":"config","log.origin":{"file.name":"config/rum.go","file.line":89},"message":"Falling back to elasticsearch output for sourcemap storage","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.510Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":144},"message":"Listening on: [::]:8200","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.511Z","log.logger":"beater","log.origin":{"file.name":"apm-server/main.go","file.line":61},"message":"creating transaction metrics aggregation with config: {Interval:1m0s MaxTransactionGroups:10000 HDRHistogramSignificantFigures:2}","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.519Z","log.logger":"beater","log.origin":{"file.name":"apm-server/main.go","file.line":61},"message":"creating transaction metrics aggregation with config: {Interval:1m0s MaxTransactionGroups:10000 HDRHistogramSignificantFigures:2}","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.521Z","log.logger":"beater","log.origin":{"file.name":"apm-server/main.go","file.line":76},"message":"creating service destinations aggregation with config: {Interval:1m0s MaxGroups:10000}","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.521Z","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":40},"message":"blocking ingestion until all preconditions are satisfied","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.522Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path / added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.523Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /config/v1/agents added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /config/v1/rum/agents added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"beater","log.origin":{"file.name":"apm-server/main.go","file.line":76},"message":"creating service destinations aggregation with config: {Interval:1m0s MaxGroups:10000}","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path / added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /config/v1/agents added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /config/v1/rum/agents added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v2/rum/events added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v3/rum/events added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v2/events added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v2/profile added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /v1/traces added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /v1/metrics added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /v1/logs added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"beater","log.origin":{"file.name":"beater/server.go","file.line":233},"message":"Starting apm-server [82f8ec0d4292f361130fc953ed8a5e7bfbe12e1c built 2022-08-15 11:30:14 +0200 +0200]. Hit CTRL-C to stop it.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v2/rum/events added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v3/rum/events added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v2/events added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /intake/v2/profile added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /v1/traces added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.524Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /v1/metrics added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"handler","log.origin":{"file.name":"api/mux.go","file.line":140},"message":"Path /v1/logs added to request handler","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/server.go","file.line":233},"message":"Starting apm-server [82f8ec0d4292f361130fc953ed8a5e7bfbe12e1c built 2022-08-15 11:30:14 +0200 +0200]. Hit CTRL-C to stop it.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":112},"message":"Stop listening on: :8200","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":88},"message":"RUM endpoints enabled!","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":91},"message":"CORS related setting `apm-server.rum.allow_origins` allows all origins. Consider more restrictive setting for production use.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":106},"message":"SSL disabled.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":88},"message":"RUM endpoints enabled!","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"warn","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":91},"message":"CORS related setting `apm-server.rum.allow_origins` allows all origins. Consider more restrictive setting for production use.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/http.go","file.line":106},"message":"SSL disabled.","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"beater/server.go","file.line":251},"message":"Server stopped","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"apm-server/main.go","file.line":189},"message":"transaction metrics aggregation stopped","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"beater","log.origin":{"file.name":"apm-server/main.go","file.line":189},"message":"service destinations aggregation stopped","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.525Z","log.logger":"esclientleg","log.origin":{"file.name":"eslegclient/connection.go","file.line":108},"message":"elasticsearch url: http://elasticsearch:9200","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.578Z","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request ok","service.name":"apm-server","url.original":"/","http.request.method":"GET","user_agent.original":"Go-http-client/1.1","source.address":"172.18.0.1","http.request.body.bytes":0,"http.request.id":"d7e8d643-b142-476a-8e0f-898c8ea1be0e","event.duration":134249,"http.response.status_code":200,"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.582Z","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request ok","service.name":"apm-server","url.original":"/","http.request.method":"GET","user_agent.original":"apm-agent-go/2.0.0","source.address":"10.11.12.13","http.request.body.bytes":0,"http.request.id":"9738e7ae-c4dc-40ba-b72d-0d4be03addfe","event.duration":97927,"http.response.status_code":200,"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.591Z","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request ok","service.name":"apm-server","url.original":"/config/v1/agents?service.name=systemtest_test","http.request.method":"GET","user_agent.original":"apm-agent-go/2.0.0","source.address":"10.11.12.13","http.request.body.bytes":0,"http.request.id":"b8d4bcac-4986-4ea1-a8e5-aef22098dae8","event.duration":77072,"http.response.status_code":200,"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.598Z","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request accepted","service.name":"apm-server","url.original":"/intake/v2/events","http.request.method":"POST","user_agent.original":"apm-agent-go/2.0.0","source.address":"10.11.12.13","http.request.id":"cdce5247-a537-47cd-a143-2d7c0316bef6","event.duration":5452366,"http.response.status_code":202,"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:20.815Z","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request ok","service.name":"apm-server","url.original":"/","http.request.method":"GET","user_agent.original":"apm-agent-go/2.0.0 (filebeat)","source.address":"127.0.0.1","http.request.body.bytes":0,"http.request.id":"650fdff2-11cb-472a-b539-968c0864cda5","event.duration":149927,"http.response.status_code":200,"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-08-15T09:58:21.422Z","log.logger":"request","log.origin":{"file.name":"middleware/log_middleware.go","file.line":61},"message":"request ok","service.name":"apm-server","url.original":"/","http.request.method":"GET","user_agent.original":"apm-agent-go/2.0.0 (metricbeat)","source.address":"127.0.0.1","http.request.body.bytes":0,"http.request.id":"ebe4c9df-7603-422e-a7d8-208908a764a0","event.duration":131856,"http.response.status_code":200,"ecs.version":"1.6.0"}
--- FAIL: TestFleetIntegrationMonitoring (20.02s)
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8868/detail/PR-8868/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Run Linux tests`
<ul>
<li>Took 11 min 15 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-8868/runs/1/steps/307/log/?start=0">here</a></li>
<li>Description: <code>./.ci/scripts/linux-test.sh</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate and publish the docker images.


- `/test windows` : Build & tests on Windows.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

apmmachine commented:
### :books: Go benchmark report

Diff with the `main` branch

```
name                                                            old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
AggregateSpan-12                                                  1.14µs ±18%    0.91µs ±16%  -20.49%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
TraceGroups-12                                                     126ns ± 5%     139ns ±12%  +10.40%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec_big_tx/1_events-12                          10.8µs ± 1%    11.4µs ± 3%   +5.46%  (p=0.008 n=5+5)
IsTraceSampled/unsampled-12                                       79.0ns ± 1%    77.8ns ± 1%   -1.53%  (p=0.016 n=5+5)

name                                                            old alloc/op   new alloc/op   delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec/100_events-12                                750kB ± 0%     750kB ± 0%   -0.01%  (p=0.032 n=5+5)

name                                                            old allocs/op  new allocs/op  delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64

```

_report generated with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat_
<!--COMMENT_GENERATED_WITH_ID_bench.diff-->

marclop commented on internal/model/modelindexer/indexer.go:
```diff
@@ -308,6 +309,17 @@ func encodeAny(v interface{}, out *fastjson.Writer) error {
+	case float64:
+		// truncate the value with int conversion to
+		// check for decimals.
+		if float64(v) == float64(int64(v)) {
+			// we cannot rely on fastjson.Marshal as it would strip
+			// the leading zero and encode the value as an "int".
+			// So we call AppendFloat directly with precision=1.
+			out.RawBytes(strconv.AppendFloat(nil, float64(v), 'f', 1, 64))
```
nit: There's no need to cast `v` to `float64` since the switch/case clause already does that for you.
```suggestion
		if v == float64(int64(v)) {
			// we cannot rely on fastjson.Marshal as it would strip
			// the leading zero and encode the value as an "int".
			// So we call AppendFloat directly with precision=1.
			out.RawBytes(strconv.AppendFloat(nil, v, 'f', 1, 64))
```

kruskall replied:
Good catch!

marclop reviewed (approved):
LGTM. It looks like the processing performance has improved as well (>-50% improvement)?

May be worth back porting this to the `8.4.` branch and hold merging until `8.4.0` is out, so we can include the fix in `8.4.1`.

axw reviewed (commented):
I don't think this is quite enough to close https://github.com/elastic/apm-server/issues/5682.

Although it's much less likely, there's still the possibility that a double value can trigger a mapping exception when relying on dynamic mapping, as double values are dynamically mapped to the "float" field type: https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-field-mapping.html

Consider this:

```
PUT /metrics
{
  "mappings": {
    "properties": {
      "float": {"type": "float"},
      "double": {"type": "double"}
    }
  }
}

# Fails
POST /metrics/_doc
{
  "float": 3.40282346638528859811704183484516925440e+39
}

# Fails
POST /metrics/_doc
{
  "dynamic": 3.40282346638528859811704183484516925440e+39
}

# Succeeds
POST /metrics/_doc
{
  "double": 3.40282346638528859811704183484516925440e+39
}
```

This is why I think we should explicitly map simple metric values to `double`. To do that, we would add another named dynamic template to the `app_metrics` data stream.

See https://github.com/elastic/apm-server/blob/1efe50c7c098b7aae754b2d4922b8aff03141a31/apmpackage/apm/data_stream/app_metrics/manifest.yml#L15-L25 and https://github.com/elastic/apm-server/blob/1efe50c7c098b7aae754b2d4922b8aff03141a31/apmpackage/apm/data_stream/app_metrics/elasticsearch/ingest_pipeline/default.yml#L22-L26

kruskall commented:
> I don't think this is quite enough to close https://github.com/elastic/apm-server/issues/5682.
> 
> Although it's much less likely, there's still the possibility that a double value can trigger a mapping exception when relying on dynamic mapping, as double values are dynamically mapped to the "float" field type: https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-field-mapping.html

Ah, I see. Good point!

I've updated the code and added the dynamic template

axw commented on apmpackage/apm/data_stream/app_metrics/manifest.yml:
```diff
@@ -23,6 +23,10 @@ elasticsearch:
                 - value_count
               default_metric: value_count
+        - doubles:
+            match_mapping_type: float
```
```suggestion
```

We should always explicitly specify the dynamic template by name, so it shouldn't be necessary to match on type.

axw commented on apmpackage/apm/data_stream/app_metrics/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -23,6 +23,8 @@ processors:
             dynamic_templates[name] = "histogram";
           } else if (metric_type == "summary") {
             dynamic_templates[name] = "summary";
+          } else if (metric_type == "counter" || metric_type == "gauge") {
```
```suggestion
          } else {
```

axw commented on apmpackage/apm/data_stream/app_metrics/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -23,6 +23,8 @@ processors:
           } else if (metric_type == "summary") {
             dynamic_templates[name] = "summary";
+          } else if (metric_type == "counter" || metric_type == "gauge") {
+            dynamic_templates[name] = "doubles";
```
```suggestion
            dynamic_templates[name] = "double";
```

Singular, for consistency with "summary" and "histogram".

axw commented on apmpackage/apm/data_stream/app_metrics/manifest.yml:
```diff
@@ -23,6 +23,10 @@ elasticsearch:
                 - sum
                 - value_count
               default_metric: value_count
+        - doubles:
```
```suggestion
        - double:
```

axw commented on internal/model/modelindexer/indexer.go:
```diff
@@ -308,6 +309,17 @@ func encodeAny(v interface{}, out *fastjson.Writer) error {
 		return encodeMap(v, out)
 	case map[string]interface{}:
 		return encodeMap(v, out)
+	case float64:
+		// truncate the value with int conversion to
+		// check for decimals.
+		if v == float64(int64(v)) {
+			// we cannot rely on fastjson.Marshal as it would strip
+			// the leading zero and encode the value as an "int".
+			// So we call AppendFloat directly with precision=1.
+			out.RawBytes(strconv.AppendFloat(nil, v, 'f', 1, 64))
+			return nil
+		}
+		return fastjson.Marshal(out, v)
```
```suggestion
```

This should not be necessary. We'll map it as a double even if it's encoded as an integer, by specifying the dynamic template by name.

kruskall replied:
The system test is failing without this line. Am I missing something ?

axw replied:
I overlooked something: I was thinking that all metrics would have an entry in the "_metric_descriptions" map which is used in the ingest processor for specifying dynamic templates. This isn't the case: we only add entries for metrics with either Type or Unit set. So we'll need some more changes:

- Always send a metric description, even if the type and unit are empty. This would enable us to identify metric fields in the ingest pipeline, and assign them the correct dynamic template. The changes would need to be made in https://github.com/elastic/apm-server/blob/a1d1cfb3e07dbf5e6ab95e5f1b7a95faddfbd374/internal/model/metricset.go#L163-L172
- Update the `internal_metrics` data stream's ingest pipeline to discard metric descriptions, since in that data stream we do not use dynamic mapping at all.

I'm open to alternative ideas.

kruskall replied:
Sorry for the long wait 🙇 

I learned quite a lot while working on this. I've pushed some changes!

axw commented on systemtest/metrics_test.go:
```diff
@@ -170,7 +170,7 @@ func TestApplicationMetrics(t *testing.T) {
 			"mapping": map[string]interface{}{
 				"c": map[string]interface{}{
-					"type": "long",
+					"type": "float",
```
This should be `double` now, right?

axw reviewed (approved):
LGTM, thanks :)

I think this one deserves a changelog entry.

axw commented on apmpackage/apm/data_stream/internal_metrics/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -9,3 +9,7 @@ processors:
       name: process_ppid
   - pipeline:
       name: client_geoip
+  - script:
+      if: ctx._metric_descriptions != null
+      source: |
+        ctx.remove("_metric_descriptions");
```
```suggestion
  - remove:
      field: _metric_descriptions
      ignore_missing: true
```

Should be a bit more efficient.

axw commented on internal/model/metricset.go:
```diff
@@ -165,9 +165,9 @@ func (me *Metricset) setFields(fields *mapStr) {
-		md.maybeSetString("type", string(sample.Type))
-		md.maybeSetString("unit", sample.Unit)
-		metricDescriptions.maybeSetMapStr(name, mapstr.M(md))
+		md.set("type", string(sample.Type))
+		md.set("unit", sample.Unit)
```
```suggestion
		md.maybeSetString("type", string(sample.Type))
		md.maybeSetString("unit", sample.Unit)
```

Can we always send the map, but only send type or unit if non-empty? This would cut down the payload sizes a little.

kruskall replied:
Good point! Changed!

kruskall commented:
/test

kruskall commented:
/test

axw reviewed (approved):
Looks great, thanks :)
